python/ray/autoscaler/v2/instance_manager/node_provider.py

```diff
@@ -1,16 +1,175 @@
+    update_id: str
+
+
+@dataclass
+class TerminateNodeError(CloudInstanceProviderError):
+    # The cloud instance id of the node that failed to terminate.
+    cloud_instance_id: CloudInstanceId
+    # From which update request the error originates.
+    update_id: str
+
+
+class ICloudInstanceProvider(ABC):
+    """
+    The interface for a cloud instance provider.
+
+    This interface is a minimal interface that should be implemented by the
+    various cloud instance providers (e.g. AWS, and etc).
+
+    The cloud instance provider is responsible for managing the cloud instances in the
+    cluster. It provides the following main functionalities:
+        - Update the cloud instances:
+            1. Given a target running set of cloud instances, may launch new ones.
+            2. Terminate existing running instances.
+        - Get the running cloud instances in the cluster.
+        - Poll the errors that happened for the updates to the cloud instance provider.
+
+    Below properties of the cloud instance provider are assumed with this interface:
+
+    1. Eventually consistent
+    The cloud instance provider is expected to be eventually consistent with the
+    cluster state. For example, when a cloud instance is request to be terminated
+    ,the provider may not immediately reflect the change in its state.
+    However, the provider is expected to eventually reflect the change in its state.
+
+    2. Asynchronous
+    The provider could also be asynchronous, where the termination/launch
+    request may not immediately return the result of the request.
+
+    3. Unique cloud instance ids
+    Cloud instance ids are expected to be unique across the cluster.
+
+    4. Idempotent updates
+    For the update APIs (e.g. update_cluster, terminate), the provider may use the
+    update ids to provide idempotency.
+
+    Usage:
+        ```
+            provider: ICloudInstanceProvider = ...
+
+            # Update the cluster with a designed shape.
+            provider.update(
+                target_running_nodes={
+                    "worker_nodes": 10,
+                    "ray_head": 1,
+                },
+                to_terminate=["node_1", "node_2"],
+                id="update_1",
+            )
+
+            # Poll the state of the cloud instance provider.
+            running = provider.get_running()
+
+            # Poll the errors
+            errors = provider.poll_errors()
+
+            # Process the state of the provider.
+            ...
+        ```
+    """
+
+    @abstractmethod
+    def get_running(self) -> Dict[CloudInstanceId, CloudInstance]:
+        """Get the running cloud instances in the cluster.
+
+        Returns:
+            A dictionary of the running cloud instances in the cluster.
+            The key is the cloud instance id, and the value is the cloud instance.
+        """
+        pass
+
+    @abstractmethod
+    def update(
+        self,
+        target_running_nodes: Dict[NodeType, int],
```

Comment:
does this include those are already running? it's a big unclear.

Reply:
yeah - this would include those running, basically a go-state for the cluster.

Reply:
if this is the case, what's the semantics if i have current state {type A: 3, type B: 2}, and target state {type A: 2, type B: 2}? looks that covers the to_terminate case as well?

Reply:
This would be a no op since the target state has been covered. I guess a more accurate term would be `required_running_nodes`

The reason it's not an equal comparison is for current state `{A: 3}` and target state `{A:2}`, while we know one of the node is extra, we don't know which one to terminate.

Reply:
why not decouple them into two apis?

Reply:
synced offline - added the `ensure_min_nodes` API.